Product Gallery: Polish Gallery in full view mode

gigitux opened the pull request:
⚠️ This PR is blocked by woocommerce/woocommerce-blocks#10823 

<!-- Please do not remove any information from this pull request. Instead, add N/A or leave blank if not applicable -->

## What

This PR is the final chunk to fix https://github.com/woocommerce/woocommerce-blocks/issues/9942.

## Why

<!-- Describe the reason for your changes. This will help the reviewer and future readers get additional context -->

This PR polishes the Gallery in full-view mode. It improves:
- [ ] Dialog UI.
- [ ] Add button on top of the dialog to close it.

## Testing Instructions

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

1. Edit the Single Product Template.
2. Add the Product Gallery block.
3. Ensure that the `Full-screen when clicked` is enabled.
4. Ensure that in the template parts list, the Product Gallery is visible.
5. Edit the Product Gallery template part.
6. Add a paragraph.
7. Save it.
8. Visit a product page.
9. Click on the main image.
10. Ensure that the modal is opened.
11. Ensure that the paragraph is visible


* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested (ie: it makes changes to tests, coding standards, docblocks, etc.). -->
* [ ] Should be tested by the development team exclusively <!-- Check this box if this PR should be tested by the development team exclusively (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

## Screenshots or screencast

<!-- Any screenshots of UI changes will be helpful to include here. Leave blank if not applicable. -->

https://github.com/woocommerce/woocommerce-blocks/assets/4463174/2ed2179f-8837-4ebb-a370-c9747736beae


## WooCommerce Visibility

<!-- Check this documentation link (../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WooCommerce core, part of the feature plugin, or experimental. -->
Required:

* [ ] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental
* [ ] N/A

## Checklist

Required:
* [ ] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [ ] This PR is assigned to a milestone.

Conditional:
* [ ] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

## Changelog
<!-- Provide a brief, descriptive summary of the changes in this PR. Include potential impacts on different parts of the product. Example: "Updated the checkout process to streamline the experience for users and reduce the number of steps." -->

> Add suggested changelog entry here.

gigitux committed:
Product Gallery: add support for On Sale Badge Block

gigitux committed:
add align support

gigitux committed:
Add E2E tests

gigitux committed:
set margin via Block Styles

gigitux committed:
Merge branch 'trunk' into 10751-product-gallery-block-add-support-for-on-sale-badge-block

gigitux committed:
disable experimental flag

gigitux committed:
Merge branch '10751-product-gallery-block-add-support-for-on-sale-badge-block' of https://github.com/woocommerce/woocommerce-blocks into 10751-product-gallery-block-add-support-for-on-sale-badge-block

gigitux committed:
Merge branch 'trunk' into 10751-product-gallery-block-add-support-for-on-sale-badge-block

gigitux committed:
add next previous block

gigitux committed:
restore support file

gigitux committed:
fix TS error

gigitux committed:
fix layout

gigitux committed:
change product

gigitux committed:
change product

gigitux committed:
Product Gallert Block: Add zoom on hover

gigitux committed:
Merge branch '9941-product-gallery-block-add-zoom-to-the-large-image-block' of https://github.com/woocommerce/woocommerce-blocks into 9941-product-gallery-block-add-zoom-to-the-large-image-block

gigitux committed:
set to true by default

gigitux committed:
Merge branch 'trunk' into 9941-product-gallery-block-add-zoom-to-the-large-image-block

gigitux committed:
remove block is already registered error

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into 9941-product-gallery-block-add-zoom-to-the-large-image-block

gigitux committed:
remove unecessary await

gigitux committed:
Merge branch 'trunk' into 9941-product-gallery-block-add-zoom-to-the-large-image-block

gigitux committed:
Improve zoom logic

Co-authored-by: Alexandre Lara <allexandrelara@gmail.com>

gigitux committed:
Merge branch 'trunk' into 9941-product-gallery-block-add-zoom-to-the-large-image-block

gigitux committed:
Product Gallery Full view mode: Add the logic to render the dedicated template

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into 9942-product-gallery-block-add-full-screen-gallery

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into 9942-product-gallery-block-add-full-screen-gallery

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into 9942-product-gallery-block-add-full-screen-gallery

gigitux committed:
use template-part instead template

gigitux committed:
add E2E tests

gigitux committed:
update selectors

gigitux committed:
add feature flag product gallery template part

gigitux committed:
fix E2E tests

gigitux committed:
remove not necessary file

gigitux committed:
polish the dialog

thealexandrelara commented on assets/js/blocks/product-gallery/frontend.tsx:
```diff
@@ -44,12 +48,23 @@ interactivityApiStore( {
+				},
+			},
+			dialog: {
+				handleClick: ( { context }: Store ) => {
```
I believe it would be better if we specify what this action is responsible for, for instance: `handleCloseButtonClick` would bring more clarity that this action is responsible for closing the dialog.

gigitux replied:
Fixed with [20b75e3](https://github.com/woocommerce/woocommerce-blocks/pull/10947/commits/20b75e302067e5f3eef1d2c4b47ab4850ce73d83)

thealexandrelara reviewed (changes requested):
Code is looking good, you've made an excellent progress on enhancing the full screen view. I have some issues that I've found while testing though:


#### Selected image is different from the image on the Full screen view

I noticed that after selecting a thumbnail image and clicking on the Large Image block, the image that opens is different from the selected one:

https://github.com/woocommerce/woocommerce-blocks/assets/20469356/da97d382-7df3-4cf9-ac9e-b4fee57f2117

It seems the context is different for each view (standard and fullscreen). I'm wondering if we will need to move the `selectedImageId` property from the `context` to the `state` in the Interactivity API store to make it global. However, this would be an issue when we have more than one Product Gallery on the same page unless we find a way to generate a unique id for each instance of the Product Gallery block 🤔 

----

#### Padding and Thumbnails block on mobile sizes

When using the tablet view, the padding is missing at the top. For the mobile view, there is a lack of padding at the top and at the bottom, and it seems there is an excessive padding at the sides.

Regarding the thumbnails they are occupying too much space on these smaller screens, but I believe it would be better to address this in a separate PR, probably the solution would be to move the thumbnails block to the bottom of the Large Image block when on mobile, what do you think?

| Laptop | Tablet | iPhone SE |
|--------|--------|--------|
| <img width="1026" alt="CleanShot 2023-09-14 at 17 07 20@2x" src="https://github.com/woocommerce/woocommerce-blocks/assets/20469356/b55a2365-efa5-41e6-bf9e-9c1f41b8c4f8"> | <img width="770" alt="CleanShot 2023-09-14 at 17 08 21@2x" src="https://github.com/woocommerce/woocommerce-blocks/assets/20469356/02d6fc5f-171c-4419-aa03-de59ca94b6ef"> | <img width="376" alt="CleanShot 2023-09-14 at 17 09 29@2x" src="https://github.com/woocommerce/woocommerce-blocks/assets/20469356/3b6f7177-40f5-4914-bf9f-b9f713d8af3c"> | 

---

#### When clicking on the Previous/Next buttons, the dialog opens

I don't know if this issue is occurring because currently we have no handler for the click on the Next/Previous buttons, but currently when clicking on them, will also trigger the action that opens the dialog:

https://github.com/woocommerce/woocommerce-blocks/assets/20469356/8df41f9b-98da-4e8b-bd00-1ad014c5169b

---

### Blank space

I'm using a 5k monitor and it seems there is too much blank space around the block:

<img width="2025" alt="CleanShot 2023-09-14 at 17 36 27@2x" src="https://github.com/woocommerce/woocommerce-blocks/assets/20469356/baaac38c-0442-40ef-9831-c750aac3ca48">

I was looking how Amazon handled this scenario, and it seems that basically they make the image occupy 100% of the dialog height and also center the image:
<img width="2021" alt="CleanShot 2023-09-14 at 17 39 27@2x" src="https://github.com/woocommerce/woocommerce-blocks/assets/20469356/32f6e7bc-bb54-4017-8ed6-bc5e141c7c5f">

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-blocks into product-gallery-full-screen-dialog-css

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-blocks into product-gallery-full-screen-dialog-css

gigitux committed:
fix: dialog show always the selected product

gigitux committed:
fix: not open the dialog when the user click on icon

gigitux committed:
rename handleClick to handleCloseButtonClick

gigitux committed:
improve style

gigitux commented:
Thanks, @thealexandrelara for your review! Sorry for the late update, but I was busy with other tasks!

### Selected image is different from the image on the Full screen view

Fixed with [ae6dc94](https://github.com/woocommerce/woocommerce-blocks/pull/10947/commits/ae6dc9469121cf6276db959008668b25dea06aee)

### When clicking on the Previous/Next buttons, the dialog opens
Fixed with [ded8a7f](https://github.com/woocommerce/woocommerce-blocks/pull/10947/commits/ded8a7f4edaecf4093e5fe6cf84be6c00976d5eb)

Regarding:
- Padding and Thumbnails block on mobile sizes
- Blank space

I created a dedicated issue: https://github.com/woocommerce/woocommerce/issues/42217. We need to find how to resolve these issues using the layout blocks of Gutenberg.

Furthermore, you will notice another style regression in full-view mode:

![image](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/f34c4af6-94bf-4d6a-82ab-a4f7d645eac7)



@danieldudzic did a great job with woocommerce/woocommerce-blocks#10867 but introduced some regression, including this one. Before fixing this CSS, I prefer to wait for @danieldudzic to fix the other regressions.

gigitux committed:
fix overlay css

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-blocks into product-gallery-full-screen-dialog-css

gigitux committed:
fix registration

gigitux commented:
Given that we fixed the function [registerBlockSingleProductTemplate](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/atomic/utils/register-block-single-product-template.ts#L26C14-L26C48) (https://github.com/woocommerce/woocommerce-blocks/pull/10978), if we register the Product Gallery with registerBlockSingleProductTemplate, the block will not be visible in the template-part. With [784e19e](https://github.com/woocommerce/woocommerce-blocks/pull/10947/commits/784e19e598f2a073f3f251192038e6f142420131), I replaced the function with `registerBlock` as a temporary solution. 
I created a dedicated issue https://github.com/woocommerce/woocommerce-blocks/issues/11027.

gigitux committed:
improve logic

gigitux committed:
improve default template

thealexandrelara reviewed (approved):
Amazing work here, @gigitux! Thank you for addressing the issues raised and for opening new ones so we can work on them in the future.

Really nice to see the dialog in sync with the Product Gallery on the template when selecting a different image 🙌 .

I tested and everything is working well (considering that we have opened issues that will be addressed in other PRs) so I'm approving this PR! 🚀

tarunvijwani commented:
Bumping the PR to the next release

gigitux committed:
Merge branch 'trunk' into product-gallery-full-screen-dialog-css
